runtime/src/hard_forks.rs

```diff
@@ -5,10 +5,12 @@ use byteorder::{ByteOrder, LittleEndian};
 use solana_sdk::clock::Slot;
 use std::ops::Add;
 
+#[frozen_abi(digest = "2vR5GzbvHHTKWA2odaV9FpSh9UdQuV4KFBbYCKYgnivz")]
 #[derive(Default, Clone, Deserialize, Serialize)]
 pub struct HardForks {
```

Comment:
`HardForks` would only be frozen because it's referenced within `Bank`. `frozen_abi` isn't able to walk the types of field members?  Ideally I think only the top-level structures that are exposed externally would need the `frozen_abi` annotation

Reply:
> `frozen_abi` isn't able to walk the types of field members?

Currently no it isn't, as the description of this PR says.

> Ideally I think only the top-level structures that are exposed externally would need the `frozen_abi` annotation

Technically, we can do that (with nightly rust). Half experimental impl is laying locally. :) Additional 1-day of my time to finish it up.

No particular preference, but there is a trade-off between more complicated impl (ie. engineering cost/risk) and accuracy (including maintenance cost).

After all, adding `frozen_abi` all around the codebase is an once-for-all event. So the current form might not be a problem in terms of maintenance.

Also, annotating only the top-level `struct`s would make it harder to notice where the digest mismatch is occurring in the future. So, annotating immediate structs adds some value for DevX in my opinion although it's tedious for the first time.

All said, my engineer's soul just says stop whining and just finish the half-baked fancy detection mechanism with latest rust's macro/template magic. But I'm just a little trainee dog. xD

Reply:
Oh, and one more consideration: with that advanced implementation, we're more locked down to the serde/bincode. But, as we're talking about ABI freeze, we've already set. ;)

Reply:
Requiring nightly rust to validate a `frozen_abi` annotation seems tolerable to me if we must.

I think my main argument for only annotating the top-level structures is simplicity for everybody else (but the macro maintainer!).  When we have a `frozen_abi` type A and forget to annotate any dependent types, we lost protection for those dependent types and thus all of A.  I can almost guarantee this will happen, because all good software engineers are lazy and if the compiler doesn't yell at them then they won't do it 100% of the time.  

As a mater of style I think the fewer `frozen_abi` annotations we have around the tree the better too.  But that's just an opinion.

> All said, my engineer's soul just says stop whining and just finish the half-baked fancy detection mechanism with latest rust's macro/template magic. But I'm just a little trainee dog. xD

lol

Reply:
Hooray, I've finally manged to detect ABI changes recursively! But I've done with a really messy implementation. ;)

Anyway, CI is green while abi tests are run by `coverage` ci job (which uses nightly) and not by `stable` ci job.

Also, I've also added necessary CI/build config plumbings to conditionally enable fronzen_abi_tests on nightly rustc.

Remaining tasks: small self-commented nits, general clean-up and some documenting as source-comments as this is a truly magic.